doc/gnpsi/gnpsi_hld.md

```diff
@@ -0,0 +1,464 @@
+
+
+
+#### Repository and Module
+
+This design introduces a SONiC application extension. The primary modification involves adding a new `sonic-gnpsi` repository. The `sonic-gnpsi` would utilize an [open source gNPSI repo on github](https://github.com/openconfig/gnpsi) to build the server code. The `sonic-gnpsi` repo would be added as a submodule into the `sonic-buildimage`
+
+By this way we ensure modularity and maintainability, making it easier to update and manage dependencies.
+
+Build dependencies include the new `sonic-gnpsi` repository, the github `openconfig/gnpsi` repository, and `bazel`.
+
+The dependency of repository is shown below
+
+![Image showing how gNPSI is integrated into SONiC repo](images/image5.png "gnpsi repos")
+
+
+
+#### Files
+
+
+
+*   `gnpsi.cc` : main file 
+*   `/db_monitor`: to monitor config db gnpsi-related events
+*   `/server`: helper to start and stop relay server and the stats thread, also write to appl_state_db
+*   `/utils`: stats util, credential util and authz logger util.
+
+
+#### Alternative
+
+![alternative design without local loopback](images/image6.png "alternative")
+
+
+**Advantages:**
+
+
+
+*   Avoids the overhead of writing and reading from the local loopback IP.
+
+**Disadvantages:**
+
+
+
+*   Introduces architectural complexities.
+*   Difficult to upstream and maintain the sFlow open-source C code.
+
+Given that the overhead of using the local loopback collector is negligible, we opted for local loopback collector solution instead of integrated processes.
+
+
+### 7. Configuration and management
+
+
+gNMI path
+
+These paths controls the enablement of the server and gRPC port for gNPSI relay server 
+
+
+
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/config/enable`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/config/port`
+
+These paths cover the stats collection for gNPSI
+
+
+
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/bytes-sent`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/packets-sent`
+*   `/system/grpc-servers/grpc-server[name=gnpsi]/clients/client[address=&lt;client_ip>][port=&lt;port_num>]/state/sample-send-error`
+
+
+Server flag
+
+
+*   `gnpsi_grpc_port`: gRPC port for the gNPSI Server
+*   `gnpsi_max_clients`: the max number of clients that can connect to gNPSI server at same time
+*   `udp_port` UDP port to read sFlow samples. Switch would configure localhost and this port as sFlow collector.
+
+
+#### 7.1 Config DB Enhancements
+
+gNPSI uses `CONFIG_DB`, `APPL_STATE_DB` and `COUNTERS_DB` 
+
+DB schema is shown as below:
+
+
+```
+CONFIG_DB:
+	"GNPSI|global": {
+"admin_state": "ENABLE"|"DISABLE",
+"port": "<port_num>"
+}
+	APPL_STATE_DB:
+		"GNPSI:global": {
+"admin_state": "ENABLE"|"DISABLE",
+"port": "<port_num>"
+}
+	COUNTERS_DB:
+		"COUNTERS:GNPSI:<remote_ip>/<port_num>": {
```

Comment:
See if you can keep last N clients and flush upon feature disable.

Reply:
Updated the doc. We clean up the stats on each client disconnection and server restart.